{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python377jvsc74a57bd033110ad40cb846f70d8671d58860389ee5a622add14023955b25793fe63fa640",
   "display_name": "Python 3.7.7 64-bit ('Samuels_Lab': conda)"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from glob import glob\n",
    "import re\n",
    "import sys\n",
    "import os\n",
    "from typing import List, Optional, Dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# setting global variables\n",
    "DAVID_ANNOTATION_DIR: str = \"../Davids_mega_annotation\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_files(file_directory: str) -> List[str]:\n",
    "    \"\"\"Function that can get all the annotated excel files from the specified directory\n",
    "    Parameters\n",
    "    __________\n",
    "    file_directory : str\n",
    "        string that list the filepath to all the xlsx files that have the MEGA array probes\n",
    "    \n",
    "    Returns\n",
    "    _______\n",
    "    List[str]\n",
    "        returns a list of strings that has all the annotation files in it\n",
    "    \"\"\"\n",
    "    cur_dir: str = os.getcwd()\n",
    "\n",
    "    annotation_file_list: List[str] = []\n",
    "\n",
    "    os.chdir(file_directory)\n",
    "\n",
    "    # gathering all the files that have and .xlsx extension and iterating through them\n",
    "    for file in glob(\"*.xlsx\"):\n",
    "\n",
    "        full_file_path: str = os.path.join(file_directory, file)\n",
    "\n",
    "        # check to make sure that the file has the format ChrXX where XX is a number\n",
    "        match_string: str = re.search(r'Chr\\d\\d', file)\n",
    "\n",
    "        if match_string:\n",
    "\n",
    "            annotation_file_list.append(full_file_path)\n",
    "    \n",
    "    os.chdir(cur_dir)\n",
    "\n",
    "    return annotation_file_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_list(gene_file: str) -> List[str]:\n",
    "    \"\"\"function to get a list of genes of interest\n",
    "    Parameter\n",
    "    _________\n",
    "    gene_file : str\n",
    "        string that contains the file path for the file contains genes of interest \n",
    "    \n",
    "    Returns\n",
    "    _______\n",
    "    List[str]\n",
    "        list of strings that has the genes from the file of interest\n",
    "    \"\"\"\n",
    "    gene_df: pd.DataFrame = pd.read_csv(gene_file, sep=\"\\t\")\n",
    "    \n",
    "    \n",
    "\n",
    "    return gene_df.gene.values.tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_variant_snps(file_list: List[str], gene_list: List[str]) -> pd.DataFrame:\n",
    "    \"\"\"Function to find the variant snps on the mega probe for a specific gene\n",
    "    Parameters\n",
    "    __________\n",
    "    file_list : List[str]\n",
    "        list of files for each chromosome mega annotation file\n",
    "    \n",
    "    gene_list : List[str]\n",
    "        list of all the genes of interest\n",
    "    \n",
    "    Returns\n",
    "    _______\n",
    "    pd.DataFrame\n",
    "        returns a pandas dataframe that has the snp information for the snp of interest\n",
    "    \"\"\"\n",
    "    # looking for the gene in each of the files\n",
    "\n",
    "    for file in file_list:\n",
    "        if re.match(r\".*Chr06.*\", file): \n",
    "\n",
    "            file_df: pd.DataFrame = pd.read_excel(file, sheet_name=\"cleaned\")\n",
    "        \n",
    "        else:\n",
    "            file_df: pd.DataFrame = pd.read_excel(file)\n",
    "\n",
    "        if \"Gene(s)\" not in file_df.columns:\n",
    "\n",
    "            print(\"expected a column called Gene(s) to be in the file\")\n",
    "            sys.exit(1)\n",
    "\n",
    "        filtered_df: pd.DataFrame = file_df.loc[file_df[\"Gene(s)\"].isin(gene_list)]\n",
    "        \n",
    "        if not filtered_df.empty:\n",
    "\n",
    "            filtered_df = filtered_df[filtered_df[\"Mutation(s)\"].str.contains(r'Missense|Nonsense', na=False)]\n",
    "            \n",
    "            if re.match(r\".*Chr06.*\", file): \n",
    "                \n",
    "                file_info_dict: Dict[str, Optional] = {\n",
    "                    \"name\":file_df.IlmnID.values.tolist(), \n",
    "                    \"RsID\":file_df[\"RS Name\"].values.tolist(),\n",
    "                    \"Chr\": file_df.Chr.values.tolist(),\n",
    "                    \"MapInfo\": file_df.MapInfo.values.tolist(),\n",
    "                    \"Alleles\": file_df.SNP.values.tolist(),\n",
    "                    \"Transcript\":None,\n",
    "                    \"Gene(s)\":file_df[\"Gene(s)\"].values.tolist(),\n",
    "                    \"In-exon\": file_df[\"In-exon\"].values.tolist(),\n",
    "                    \"Mutation(s)\": file_df[\"Mutation(s)\"].values.tolist()\n",
    "                    }\n",
    "                \n",
    "                file_info_df: pd.DataFrame = pd.DataFrame.from_dict(file_info_dict)\n",
    "\n",
    "                file_info_df.to_csv(\"mega_probes.txt\", sep=\"\\t\", mode=\"a+\", index=False)\n",
    "\n",
    "            else:\n",
    "\n",
    "                filtered_df.to_csv(\"mega_probes.txt\",sep=\"\\t\", mode=\"a+\", index=False)\n",
    "                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_list: List[str] = get_files(DAVID_ANNOTATION_DIR)\n",
    "\n",
    "gene_list: List[str] = get_gene_list(\"./retinopathy_gene_targets.txt\")\n",
    "\n",
    "find_variant_snps(file_list, gene_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ]
}